### PR TITLE
feat: NGS-bits SampleSimilarity

### PR DESCRIPTION
**Pull request by tdayris**
Adds a wrapper for [ngs-bits SampleSimilarity](https://github.com/imgag/ngs-bits/blob/master/doc/tools/SampleSimilarity/index.md). There are a lot of subcommands for ngs-bits. Many of them requires a connection to a mysql database ([NGSD](https://github.com/imgag/ngs-bits/blob/master/doc/install_ngsd.md)), which I don't know how to deploy smoothly through a snakemake wrapper.

I'll wrap some other subcommands soon.

### QC

* [X] I confirm that I have followed the [documentation for contributing to `snakemake-wrappers`](https://snakemake-wrappers.readthedocs.io/en/stable/contributing.html).

While the contributions guidelines are more extensive, please particularly ensure that:
* [X] `test.py` was updated to call any added or updated example rules in a `Snakefile`
* [X] `input:` and `output:` file paths in the rules can be chosen arbitrarily
* [X] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`)
* [X] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to 
* [X] the `meta.yaml` contains a link to the documentation of the respective tool or command under `url:`
* [X] conda environments use a minimal amount of channels and packages, in recommended ordering

## Release Notes

- **New Features**
	- Introduced `environment.linux-64.pin.txt` and `environment.yaml` files for managing package dependencies in a Conda environment.
	- Added a new metadata file `meta.yaml` for the NGS-bits SampleSimilarity tool, detailing input/output specifications and documentation links.
	- Added two new VCF files (`a.vcf` and `b.vcf`) containing genomic variant data for testing purposes.
	- Created a `wrapper.py` file to facilitate execution of the SampleSimilarity tool within a Snakemake workflow.

- **Tests**
	- Implemented a new test function for the SampleSimilarity wrapper to ensure functionality and correctness.

### CODE REVIEW

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    stderr=True,
+)
+extra = snakemake.params.get("extra", "")
+ref = snakemake.input.get("ref")
+if ref:
+    extra += f" -ref {ref} "
+
+roi = snakemake.input.get("roi")
+if roi:
+    extra += f" -roi {roi} "
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngsbits/samplesimilarity/wrapper.py]: Solved.

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+input_files = snakemake.input.get("samples")
+if all(str(i).endswith(("vcf", "vcf.gz")) for i in input_files):
+    extra += " -mode vcf "
+elif str(str(i).endswith(("sam", "bam", "cram")) for i in input_files):
+    extra += " -mode bam "
+else:
+    extra += " -mode gsvar "
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngsbits/samplesimilarity/wrapper.py]: Fixed.

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+shell("SampleSimilarity {extra} -in {input_files} -out {snakemake.output} {log}")
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngsbits/samplesimilarity/wrapper.py]: Input files can be multiple and therefore no quotes should be added.

File: bio/ngsbits/samplesimilarity/test/Snakefile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        # roi="", # Optional path to regions of interest (bed)
+    output:
+        "similarity.tsv",
+    threads: 1
+    log:
+        "samplesimilarity.log",
+    params:
+        extra="-build hg19",
+    wrapper:
+        "master/bio/ngsbits/samplesimilarity"
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngsbits/samplesimilarity/test/Snakefile:14]: Declined.

File: bio/ngsbits/samplesimilarity/environment.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+channels:
+  - conda-forge
+  - bioconda
+  - nodefaults
+dependencies:
+  - ngs-bits=2022_12
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/environment.yaml:6]: latest version is `2024_11`, no?

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from snakemake.shell import shell
+
+log = snakemake.log_fmt_shell(
+    stdout=True,
+    stderr=True,
+)
+extra = snakemake.params.get("extra", "")
+ref = snakemake.input.get("ref")
+if ref:
+    extra += f" -ref '{ref}' "
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/wrapper.py:19]:
```suggestion
    extra += f" -ref {ref:q}"
```

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+roi = snakemake.input.get("roi")
+if roi:
+    extra += f" -roi '{roi}' "
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/wrapper.py:23]:
```suggestion
    extra += f" -roi {roi:q}"
```

File: bio/ngsbits/samplesimilarity/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+name: NGS-bits SampleSimilarity
+url: https://github.com/imgag/ngs-bits/blob/master/doc/tools/SampleSimilarity/index.md
+authors:
+  - Thibault Dayris
+input:
+  - samples: list of paths to vcf/vcf.gz files, or list of paths to bam/sam/cram files
+  - ref: Optional path to reference genome index file (FAI). Required for CRAM input.
+  - roi: Optional path to regions of interest (BED).
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/meta.yaml:8]: Maybe rename to `regions` so it is consistent with other wrappers?

File: bio/ngsbits/samplesimilarity/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+name: NGS-bits SampleSimilarity
+url: https://github.com/imgag/ngs-bits/blob/master/doc/tools/SampleSimilarity/index.md
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/meta.yaml:2]:
```suggestion
description: Calculate several metrics that measure sample similarity.
url: https://github.com/imgag/ngs-bits/blob/master/doc/tools/SampleSimilarity/index.md
```

File: bio/ngsbits/samplesimilarity/wrapper.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+elif all(str(i).endswith(("sam", "bam", "cram")) for i in input_files):
+    extra += " -mode bam "
+else:
+    extra += " -mode gsvar "
+
+shell(
+    "SampleSimilarity {extra} "
+    "-in {input_files} "
+    "-out '{snakemake.output}' "
+    "{log}"
```
Comment (fgvieira) [inline bio/ngsbits/samplesimilarity/wrapper.py:37]:
```suggestion
    extra += f" -ref {ref:q}"

roi = snakemake.input.get("regions")
if roi:
    extra += f" -roi {roi:q}"

input_files = snakemake.input.get("samples")
if all(str(i).endswith((".vcf", ".vcf.gz")) for i in input_files):
    extra += " -mode vcf"
elif all(str(i).endswith((".sam", ".bam", ".cram")) for i in input_files):
    extra += " -mode bam"
else:
    extra += " -mode gsvar"

shell(
    "SampleSimilarity"
    " -in {input_files}"
    " {extra}"
    " -out {snakemake.output:q}"
    " {log}"
```

**Review by fgvieira — APPROVED** [review]

**Pull request merged**